components/tenant-mgt/org.wso2.carbon.tenant.mgt/src/main/java/org/wso2/carbon/tenant/mgt/services/TenantMgtImpl.java

```diff
@@ -0,0 +1,493 @@
+import org.wso2.carbon.stratos.common.exception.TenantManagementServerException;
+import org.wso2.carbon.stratos.common.exception.TenantMgtException;
+import org.wso2.carbon.stratos.common.util.CommonUtil;
+import org.wso2.carbon.tenant.mgt.core.TenantPersistor;
+import org.wso2.carbon.tenant.mgt.internal.TenantMgtServiceComponent;
+import org.wso2.carbon.tenant.mgt.util.TenantMgtUtil;
+import org.wso2.carbon.user.api.RealmConfiguration;
+import org.wso2.carbon.user.api.TenantMgtConfiguration;
+import org.wso2.carbon.user.api.UserRealm;
+import org.wso2.carbon.user.core.UserCoreConstants;
+import org.wso2.carbon.user.core.UserStoreException;
+import org.wso2.carbon.user.core.UserStoreManager;
+import org.wso2.carbon.user.core.common.User;
+import org.wso2.carbon.user.core.config.multitenancy.MultiTenantRealmConfigBuilder;
+import org.wso2.carbon.user.core.constants.UserCoreClaimConstants;
+import org.wso2.carbon.user.core.service.RealmService;
+import org.wso2.carbon.user.core.tenant.Tenant;
+import org.wso2.carbon.user.core.tenant.TenantManager;
+import org.wso2.carbon.user.core.tenant.TenantSearchResult;
+import org.wso2.carbon.utils.multitenancy.MultitenantConstants;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+
+import static org.wso2.carbon.stratos.common.constants.TenantConstants.ErrorMessage.ERROR_CODE_INVALID_OFFSET;
+import static org.wso2.carbon.tenant.mgt.util.TenantMgtUtil.initializeTenantInfoBean;
+
+/**
+ * Default implementation of {@link TenantMgtService} interface.
+ */
+public class TenantMgtImpl implements TenantMgtService {
+
+    private static final Log log = LogFactory.getLog(TenantMgtImpl.class);
+    private static final String DEFAULT_SORT_BY = "UM_DOMAIN_NAME";
+    private static final String DESC_SORT_ORDER = "DESC";
+    private static final String ASC_SORT_ORDER = "ASC";
+    public static final String DOMAIN_NAME = "domainName";
+
+    public String addTenant(Tenant tenant) throws TenantMgtException {
+
+        String tenantDomain = tenant.getDomain();
+        int tenantId;
+
+        validateInputs(tenant);
+        try {
+            // Set a thread local variable to identify the operations triggered for a tenant admin user.
+            TenantMgtUtil.setTenantAdminCreationOperation(true);
+
+            addAttributeValues(tenant);
+            createTenant(tenant);
+            addTenantAdminUser(tenant);
+            tenantId = tenant.getId();
+            addClaimsToUserStore(tenant, tenantDomain, tenantId);
+
+            TenantInfoBean tenantInfoBean = initializeTenantInfoBean(tenantId, tenant);
+            notifyTenantAddition(tenantInfoBean);
+
+            // For the super tenant tenant creation, tenants are always activated as they are created.
+            TenantMgtUtil.activateTenantInitially(tenantInfoBean, tenantId);
+        } catch (Exception e) {
+            if (e instanceof TenantManagementClientException) {
```

Comment:
Why don't we check TenantManagementException  instead of TenantManagementClientException?